src/driver-steelseries.c

```diff
@@ -440,98 +508,84 @@ steelseries_write_led_v2(struct ratbag_led *led)
 {
 	struct ratbag_device *device = led->profile->device;
 	uint8_t buf[STEELSERIES_REPORT_SIZE] = {0};
-	uint16_t duration, min_duration;
 	int ret;
 
-	buf[0] = STEELSERIES_ID_LED;
-	buf[2] = led->index;
+	struct steelseries_cycle cycle;
+	struct steelseries_point point1;
+	struct steelseries_point point2;
+	struct steelseries_point point3;
+	struct steelseries_point point4;
 
-	/* not sure if these two are needed */
-	buf[15] = 0x01;
-	buf[17] = 0x01;
+	create_cycle(&cycle);
+	cycle.led_id = led->index;
 
 	switch(led->mode) {
 	case RATBAG_LED_OFF:
-		buf[19] = 0x01;
-		buf[27] = 0x01;
-		buf[28] = buf[31] = 0x00;
-		buf[29] = buf[32] = 0x00;
-		buf[30] = buf[33] = 0x00;
-		/* not sure why the duration is set for the steady color or why
-		   it is different for the two LEDs */
-		duration = led->index == 0 ? 10000 : 5000; /* 0x1027 or 0x8813 */
+		point1.color.red = 0x00;
+		point1.color.green = 0x00;
+		point1.color.blue = 0x00;
+		point1.pos = 0x00;
+
+		list_append(&cycle.points, &point1.link);
 		break;
 	case RATBAG_LED_ON:
-		buf[19] = 0x01;
-		buf[27] = 0x01;
-		buf[28] = buf[31] = led->color.red;
-		buf[29] = buf[32] = led->color.green;
-		buf[30] = buf[33] = led->color.blue;
-
-		/* not sure why the duration is set for the steady color or why
-		   it is different for the two LEDs */
-		duration = led->index == 0 ? 10000 : 5000; /* 0x1027 or 0x8813 */
+		point1.color = led->color;
+
+		list_append(&cycle.points, &point1.link);
 		break;
 	case RATBAG_LED_CYCLE:
-		buf[27] = 0x04; /* number of steps in cycle */
-
-		/* start color */
-		buf[28] = buf[31] = 0xFF;
-		buf[29] = buf[32] = 0x00;
-		buf[30] = buf[33] = 0x00;
-
-		/* Cycle to green */
-		buf[35] = 0x00;
-		buf[36] = 0xFF;
-		buf[37] = 0x00;
-		buf[38] = 0x54; /* normalized time share of animation */
-
-		/* Cycle to blue */
-		buf[39] = 0x00;
-		buf[40] = 0x00;
-		buf[41] = 0xFF;
-		buf[42] = 0x54; /* normalized time share of animation */
-
-		/* Cycle to red */
-		buf[43] = 0xFF;
-		buf[44] = 0x00;
-		buf[45] = 0x00;
-		buf[46] = 0x56; /* normalized time share of animation */
-
-		duration = led->ms;
+		point1.color.red = 0xFF;
+		point1.color.green = 0x00;
+		point1.color.blue = 0x00;
+		point1.pos = 0x00;
+
+		point2.color.red = 0x00;
+		point2.color.green = 0xFF;
+		point2.color.blue = 0x00;
+		point2.pos = 0x54;
+
+		point3.color.red = 0x00;
+		point3.color.green = 0x00;
+		point3.color.blue = 0xFF;
+		point3.pos = 0x54;
```

Comment:
is this a typo? should this be `0x55`? seems strange that we have `0x00 0x54 0x54 0x56`

[later]
Ok, this is a time-share. Why is it called pos? and if it's a normalized time-share why is it in hex? The three add up to 254, is this normalized to 255? If so, one is missing ;)

Reply:
I don't really know, this is how the official driver does it I think. I didn't touch any of the values, I just changed the implementation to use `construct_cycle_buffer()`.

I called it pos as in the position in the cycle. I can change it if you want. It made more sense to me to think of it as the position offset rather than a delay, since the the actual duration of the cycle is variable.

Reply:
calling it pos would be ok if there's a comment that explains what the position is that it refers to. This is hw-specific stuff and it's reverse-engineered, so always err on the side of too many comments that explain what you wanted to do because there's always the chance of misinterpretation. It's easy to fix something that's clearly wrong than something where you have to guess what the original intention was. Compare the fictitious example of:
```
buf[25] = 0xac; /* checksum for buf[0 - 10] */
````
with just an assignment of 
```
buf[25] = 0xac;
```
In the former you can immediately change it once you figure out that it's not the checksum (or from a different slice). The latter you're always left wondering *why* the original author set this buffer to 0xac and whether there was some intrinsic need for it.